src/Stratis.Bitcoin/Base/ConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,461 @@
+        private HeadersPayload ConstructHeadersPayload(BlockLocator locator, uint256 hashStop)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(locator), locator, nameof(hashStop), hashStop);
+
+            ChainedHeader fork = this.chain.FindFork(locator);
+
+            if (fork == null)
+            {
+                this.logger.LogTrace("(-)[INVALID_LOCATOR]:null");
+                return null;
+            }
+
+            var headers = new HeadersPayload();
+
+            foreach (ChainedHeader header in this.chain.EnumerateToTip(fork).Skip(1))
+            {
+                headers.Headers.Add(header.Header);
+
+                if ((header.HashBlock == hashStop) || (headers.Headers.Count == MaxItemsPerHeadersMessage))
+                    break;
+            }
+
+            this.logger.LogTrace("(-):'{0}'", headers);
+            return headers;
+        }
+
+        /// <summary>
+        /// Processes "headers" message received from the peer.
+        /// </summary>
+        /// <param name="peer">Peer from which the message was received.</param>
+        /// <param name="headersPayload">Payload of "headers" message to process.</param>
+        /// <remarks>
+        /// "headers" message is sent in response to "getheaders" message or it is solicited
+        /// by the peer when a new block is validated (unless in IBD).
+        /// <para>
+        /// When we receive "headers" message from the peer, we can adjust our knowledge
+        /// of the peer's view of the chain. We update its pending tip, which represents
+        /// the tip of the best chain we think the peer has.
+        /// </para>
+        /// </remarks>
+        private async Task ProcessHeadersAsync(INetworkPeer peer, HeadersPayload headersPayload)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}:'{3}')", nameof(peer), peer.RemoteSocketEndpoint, nameof(headersPayload), headersPayload);
+
+            List<BlockHeader> headers = headersPayload.Headers;
+
+            if (headers.Count == 0)
+            {
+                this.logger.LogTrace("Headers payload with no headers was received. Assuming we're synced with the peer.");
+                this.logger.LogTrace("(-)[NO_HEADERS]");
+                return;
+            }
+
+            if (headers.Count > MaxItemsPerHeadersMessage)
+            {
+                this.logger.LogTrace("Headers payload with {0} headers was received. Protocol violation. Banning the peer.", headers.Count);
+                this.peerBanning.BanAndDisconnectPeer(peer.PeerEndPoint, this.connectionManager.ConnectionSettings.BanTimeSeconds, "Protocol violation.");
+
+                this.logger.LogTrace("(-)[TOO_MANY_HEADERS]");
+                return;
+            }
+
+            // Check headers for consecutiveness.
+            for (int i = 1; i < headers.Count; i++)
```

Comment:
This should still go in a separate method as it should be unit tested. Looks like a rule.

Reply:
```
if(!CheckHeadersForConsecutiveness())
    return;
```

Reply:
Fix this and we are golden 👍

Reply:
How would moving the logic to a separate method violate single responsibility?

Hmm you do however raise an interesting point re banning and disconnection... You can actually move both those `if` that potentially ban and disconnect the peer to method:

```
if(CheckIfPeerShouldBeBanned(headers))
    return;
```

Reply:
Good idea. Moved checks that may load to peer banning to a separated method.